File: app/src/main/java/com/tama/command/CommandStep.java
package com.tama.command;

import com.tama.core.Tile;

import com.tama.core.Direction;
import com.tama.core.Pet;
import com.tama.core.World;

class CommandStep extends Command
{
    Direction dir;

    public CommandStep(Direction dir)
    {
        this.dir = dir;
    }

    public void start(Pet pet, World world)
    {
        pet.setDir(dir);
        Tile tile = world.getTile(pet.loc.x + dir.x, pet.loc.y + dir.y);
        if (pet.canMoveOnto(tile))
        {
            world.removeThing(pet);
            world.add(pet, pet.loc.x + dir.x, pet.loc.y + dir.y);
            updater = this::doing;
            return;
        }
        state = CommandState.failed;
    }

    public void doing(Pet pet, World world)
    {

    }
}

File: app/src/main/java/com/tama/command/CommandStep.kt
package com.tama.command

import android.util.Log
import com.tama.core.DisplayAdapter
import com.tama.thing.Direction
import com.tama.thing.Pet
import com.tama.core.World
import com.tama.core.WorldObject

public class CommandStep(var dir: Direction) : Command()
{

    public override fun start(pet: Pet, world: World)
    {
        super.start(pet, world)
        Log.d(javaClass.canonicalName, "stepping " + dir.x + " " + dir.y);
        pet.setDir(dir)
        val tile = world.getTile(pet.loc.x + dir.x, pet.loc.y + dir.y)
        if (pet.canMoveOnto(tile))
        {
            pet.setMovementPose(Pet.Movement.walk);
            world.removeThing(pet)
            world.add(pet, pet.loc.x + dir.x, pet.loc.y + dir.y)
            pet.loc.xoff = -dir.x * 100;
            pet.loc.yoff = -dir.y * 100;
            update = ::doing
            return
        }
        Log.d(javaClass.canonicalName, "failed step");
        state = CommandState.failed
    }

    public override fun doing(pet: Pet, world: World)
    {
        when
        {
            pet.loc.xoff != 0 ->
            {
                pet.loc.xoff += dir.x;
            }
            pet.loc.yoff != 0 ->
            {
                pet.loc.yoff += dir.y;
            }
            else              ->
            {
                pet.setMovementPose(Pet.Movement.stand);
                state = CommandState.complete;
            }
        }
    }

    override fun hardCancel()
    {
        super.hardCancel()
        actor?.loc?.xoff = 0;
        actor?.loc?.yoff = 0;
    }

    override fun draw(d: DisplayAdapter)
    {

    }
}